CPP: Make LambdaCaptures Elements

geoffw0 opened the pull request:
This is a fixed up part of something I was working on in 2017.  A `LambdaCapture` has a `Location` so there's no reason for it not to be an `Element` like everything else.

When someone is prepared to merge this I will write a (trivial) upgrade script for it.

zlaski-semmle commented:
LGTM, but can you elaborate on the script you plan to write?

jbj commented on cpp/ql/test/library-tests/lambdas/captures/elements.expected:
```diff
@@ -56,11 +58,13 @@
 | captures.cpp:10:7:10:15 | ExprStmt |
 | captures.cpp:10:9:10:9 | definition of x |
 | captures.cpp:10:9:10:9 | x |
+| captures.cpp:10:9:10:9 | x |
```
I don't understand why a second copy of something that stringifies as `x` appeared as a result of this change. This is on line 10, where there's no lambda capture, so why is there a new element here? Is something funny going on with the locations of these lambda captures?

geoffw0 replied:
This particular test doesn't stringify, it outputs each `Element` and in this case there are two - a `Field` and a `LambdaCapture`.  Previously we wouldn't have output the `LambdaCapture` but only because it wasn't an `Element`.

I'm not a fan of tests that output all `Element`s, I could perhaps modify it to only output a fixed list of things (perhaps `Variable`, `VariableDeclarationEntry`, `ExprStmt` and `LambdaCapture`) along with what they are?

jbj replied:
I think my confusion is that the capture and definition of `x` ought to belong on line 9, which reads `[=] {`, rather than line 10, which reads `b(x + 1);`. But maybe that's just an off-by-one somewhere. It's certainly not related to this PR.

Now that we have `getCanonicalQLClass`, it should be easier to write those tests that stringify a set of class names. But that's not for this PR.

geoffw0 commented:
> can you elaborate on the script you plan to write?

By 'trivial', I mean that the upgrade script won't do anything, but it's required as the `.dbscheme` has changed and won't be considered compatible otherwise (see https://wiki.semmle.com/display/IN/Creating+an+ODASA+Database+Upgrade+Script for some background on this topic).  I don't want to write it until the last minute because, if somebody else changes the `.dbscheme` before it is merged, it would have to be redone.

jbj commented:
I'm happy to merge this once there's an upgrade script. Remember that upgrade scripts go in this repo now. I've [asked on Slack](https://semmle.slack.com/archives/C0UGEJUDV/p1565361054236700) whether the internal repo also needs to be updated.

geoffw0 commented:
Added an upgrade script (https://git.semmle.com/Semmle/code/pull/33623) and updated stats (here).  I haven't done this for a while so I followed the steps on the wiki, hopefully I've got it right.

geoffw0 commented:
... pretty sure it's wrong.  I'll have to fix this another day.

geoffw0 commented:
Upgrade script moved to here and updated.  https://git.semmle.com/Semmle/code/pull/33623 should still be merged at the same time as it contains the LegacyLanguage.java change, which I assume is still needed.

geoffw0 commented:
Upgrade script is still valid.  Please review and merge.

geoffw0 commented:
Updated.
